src/core/implication.rs

```diff
@@ -0,0 +1,88 @@
+use std::fmt::Display;
+
+use itertools::Itertools;
+
+use crate::utils::{Binding, SExp};
+
+use super::{expr::EvalBool, OrderConstraint, Expr};
+
+/// A type representing the expression a => b
+#[derive(Clone)]
+pub struct Implication<T> {
+    guard: Option<OrderConstraint<T>>,
+    cons: OrderConstraint<T>
+}
+
+impl<T> Implication<T> {
+    pub fn new(guard: OrderConstraint<T>, cons: OrderConstraint<T>) -> Self {
```

Comment:
I would rename `new` and `new_opt` to be `fact` and `implication` since a non-guarded assumptions is fact that should always be true